packages/effect/src/internal/stream.ts

```diff
@@ -5675,6 +5676,90 @@ export const scoped = <A, E, R>(
 ): Stream.Stream<A, E, Exclude<R, Scope.Scope>> =>
   new StreamImpl(channel.ensuring(channel.scoped(pipe(effect, Effect.map(Chunk.of))), Effect.void))
 
+/** @internal */
+export const share = dual<
+  <A, E>(config: {
+    readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+  }) => <R>(
+    self: Stream.Stream<A, E, R>
+  ) => Effect.Effect<Stream.Stream<A, E>, never, R | Scope.Scope>,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ) => Effect.Effect<Stream.Stream<A, E>, never, R | Scope.Scope>
+>(
+  2,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ) =>
+    Effect.gen(function*() {
+      const connector = yield* (config.connector ?? PubSub.bounded<Take.Take<A, E>>(16))
+      yield* Effect.forkScoped(runIntoPubSub(self, connector))
+      return flattenTake(fromPubSub(connector))
+    })
+)
+
+/** @internal */
+export const shareRefCount = dual<
+  <A, E>(config: {
+    readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+  }) => <R>(
+    self: Stream.Stream<A, E, R>
+  ) => Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope>,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ) => Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope>
+>(
+  2,
+  <A, E, R>(
+    self: Stream.Stream<A, E, R>,
+    config: {
+      readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
+    }
+  ): Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope> => {
+    return Effect.gen(function*() {
+      let refCount = 0
+      let connector: PubSub.PubSub<Take.Take<A, E>> | null = null
+      let fiber: Fiber.RuntimeFiber<void> | null = null
+      const scope = yield* Effect.scope
+      const context = yield* Effect.context<R>()
+      const sem = yield* Effect.makeSemaphore(1)
+
+      return Effect.gen(function*() {
+        refCount++
+        connector ??= yield* (config.connector ?? PubSub.bounded<Take.Take<A, E>>(16))
+        fiber ??= yield* runIntoPubSub(self, connector).pipe(
+          Effect.locally(FiberRef.currentContext, context),
+          Effect.forkIn(scope)
```

Comment:
```ts
declare const stream: Stream<number, never, Foo>
Effect.gen(function* () {
  const shared = yield* sharedRefCount(stream)
  yield* doSomethingWith(shared).pipe(
    Effect.provide(Foo1Layer),
    Effect.forkScoped,
  )
  yield* doSomethingWith(shared).pipe(
    Effect.provide(Foo2Layer),
    Effect.forkScoped
  )
})

```
@tim-smart if I understand correctly, both instances would use the `Foo1` version. No?

Reply:
They would use whatever was supplied to the `sharedRefCount` effect.

Reply:
despite this effect being ran as part of the stream evaluation?  How?